google/cloud/bigquery_storage/stream_writer.py

```diff
@@ -0,0 +1,334 @@
+import sys
+
+from google.api_core.retry import Retry
+from google.protobuf import descriptor_pb2  # noqa: F401
+
+from google.cloud.bigquery_storage_v1 import types
+from google.cloud.bigquery_storage import base
+from google.cloud.bigquery_storage import constants
+from typing import Optional, Dict, Any
+
+BUILD_SPECIFIC_GCLOUD_PROJECT = "BUILD_SPECIFIC_GCLOUD_PROJECT"
+
+
+class WriteSession:
+    """A manager object that keeps track of write streams and their metadata."""
+
+    streams: Dict[str, types.ProtoSchema] = {}
+    connections: Dict[str, "WriteStream"] = {}
+    project = (
+        None
+        if not os.getenv(BUILD_SPECIFIC_GCLOUD_PROJECT)
+        else os.getenv(BUILD_SPECIFIC_GCLOUD_PROJECT)
+    )
+    enable_connection_pool = False
+    min_connections = 1
+    max_connections = 20
+
+    def __init__(
+        self,
+        enable_connection_pool: bool = False,
+        min_connections: Optional[int] = 1,
+        max_connections: Optional[int] = 20,
+    ):
+        """Constructs the WriteSession object.
+
+        Args:
+            enable_connection_pool (bool):
+                Optional. Defaults to False. This needs to be set to True
+                in order to set the min_connections and max_connections parameters.
+            min_connections (int):
+                Optional. Defaults to 1. Sets the minimum number of connections in the connection pool.
+            max_connections (int):
+                Optional. Defaults to 20. Sets the maximum number of connections in the connection pool.
+        """
+        self.client = base.global_base.client
+        self.enable_connection_pool = enable_connection_pool
+        self.min_connections = min_connections
+        self.max_connections = max_connections
+
+    def add_stream(
+        self,
+        table: str,
+        dataset: str,
+        stream_type: Optional[types.WriteStream.Type] = None,
+    ):
+        """Creates and adds a write stream and its metadata to the write session.
+
+        Args:
+            table (str):
+                Required. The name of the table that is written to.
+            dataset (str):
+                Required. The name of the dataset that contains the table.
+            stream_type (types.WriteStream.Type):
+                Optional. If not set, the default stream will be used.
+
+        Returns:
+            stream (WriteStream): The newly created WriteStream object.
+        """
+        stream = WriteStream(table=table, dataset=dataset, stream_type=stream_type)
+
+        return stream
+
+
+class WriteStream(WriteSession, Retry):
+    def __init__(
+        self,
+        table: str,
+        dataset: str,
+        stream_type: Optional[types.WriteStream.Type] = None,
+        enable_connection_pool: bool = False,
+        min_connections: Optional[int] = 1,
+        max_connections: Optional[int] = 20,
+        project: Optional[str] = None,
+    ):
+        """Constructs the WriteStream object.
+
+        Args:
```

Comment:
You should likely allow the user to pass in a stream id as an optional argument, to support users invoking CreateWriteStream directly.

Also, table should likely become optional when you do so.